File: src/music/queue.js
/**
 * Queue with loop modes (none / track / queue), shuffle, remove, and move.
 */

export const LoopMode = Object.freeze({
  NONE:  'none',
  TRACK: 'track',
  QUEUE: 'queue',
});

export class Queue {
  constructor(maxSize = 500) {
    this._tracks  = [];
    this._history = [];
    this.loopMode = LoopMode.NONE;
    this.maxSize  = maxSize;
  }

  // ── Properties ─────────────────────────────────────────────────────────────

  get tracks()          { return [...this._tracks]; }
  get size()            { return this._tracks.length; }
  get isEmpty()         { return this._tracks.length === 0; }
  get totalDurationMs() { return this._tracks.reduce((s, t) => s + (t.info?.length ?? 0), 0); }
  get history()         { return [...this._history]; }

  // ── Adding tracks ──────────────────────────────────────────────────────────

  add(track) {
    if (this._tracks.length >= this.maxSize) return false;
    this._tracks.push(track);
    return true;
  }

  addMany(tracks) {
    let added = 0;
    for (const track of tracks) {
      if (!this.add(track)) break;
      added++;
    }
    return added;
  }

  // ── Consuming tracks ───────────────────────────────────────────────────────

  /**
   * Dequeue the next track, honouring the current loop mode.
   * @param {import('../lavalink/models.js').Track|null} current
   * @returns {import('../lavalink/models.js').Track|null}
   */
  getNext(current = null) {
    if (this.loopMode === LoopMode.TRACK && current !== null) {
      return current;
    }

    if (this._tracks.length === 0) return null;

    const track = this._tracks.shift();

    if (current !== null) {
      this._history.push(current);
      if (this._history.length > 50) this._history.shift();
    }

    if (this.loopMode === LoopMode.QUEUE) {
      this._tracks.push(track);
    }

    return track;
  }

  remove(index) {
    if (index >= 0 && index < this._tracks.length) {
      return this._tracks.splice(index, 1)[0];
    }
    return null;
  }

  clear() {
    this._tracks.length = 0;
  }

  // ── Reordering ─────────────────────────────────────────────────────────────

  shuffle() {
    for (let i = this._tracks.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this._tracks[i], this._tracks[j]] = [this._tracks[j], this._tracks[i]];
    }
  }

  move(fromIndex, toIndex) {
    const len = this._tracks.length;
    if (fromIndex < 0 || fromIndex >= len || toIndex < 0 || toIndex >= len) return false;
    const [track] = this._tracks.splice(fromIndex, 1);
    this._tracks.splice(toIndex, 0, track);
    return true;
  }
}

File: music/queue.py
"""
Queue with loop modes (none / track / queue), shuffle, remove, and move.
"""
from __future__ import annotations

import random
from enum import Enum
from typing import List, Optional

from lavalink.models import Track


class LoopMode(Enum):
    NONE = "none"
    TRACK = "track"
    QUEUE = "queue"


class Queue:
    def __init__(self, max_size: int = 500) -> None:
        self._tracks: List[Track] = []
        self._history: List[Track] = []
        self.loop_mode: LoopMode = LoopMode.NONE
        self.max_size: int = max_size

    # ──────────────────────────────────────────────────────────────────────────
    # Properties
    # ──────────────────────────────────────────────────────────────────────────

    @property
    def tracks(self) -> List[Track]:
        return list(self._tracks)

    @property
    def size(self) -> int:
        return len(self._tracks)

    @property
    def is_empty(self) -> bool:
        return not self._tracks

    @property
    def total_duration_ms(self) -> int:
        return sum(t.info.length for t in self._tracks)

    @property
    def history(self) -> List[Track]:
        return list(self._history)

    # ──────────────────────────────────────────────────────────────────────────
    # Adding tracks
    # ──────────────────────────────────────────────────────────────────────────

    def add(self, track: Track) -> bool:
        if len(self._tracks) >= self.max_size:
            return False
        self._tracks.append(track)
        return True

    def add_many(self, tracks: List[Track]) -> int:
        added = 0
        for track in tracks:
            if self.add(track):
                added += 1
            else:
                break
        return added

    # ──────────────────────────────────────────────────────────────────────────
    # Consuming tracks
    # ──────────────────────────────────────────────────────────────────────────

    def get_next(self, current: Optional[Track] = None) -> Optional[Track]:
        """
        Dequeue the next track, honouring the current loop mode.
        `current` is the track that just finished; it is pushed to history
        (and re-queued if QUEUE loop is on).
        """
        if self.loop_mode == LoopMode.TRACK and current is not None:
            return current

        if not self._tracks:
            return None

        track = self._tracks.pop(0)

        if current is not None:
            self._history.append(current)
            if len(self._history) > 50:
                self._history.pop(0)

        if self.loop_mode == LoopMode.QUEUE:
            self._tracks.append(track)

        return track

    def remove(self, index: int) -> Optional[Track]:
        if 0 <= index < len(self._tracks):
            return self._tracks.pop(index)
        return None

    def clear(self) -> None:
        self._tracks.clear()

    # ──────────────────────────────────────────────────────────────────────────
    # Reordering
    # ──────────────────────────────────────────────────────────────────────────

    def shuffle(self) -> None:
        random.shuffle(self._tracks)

    def move(self, from_index: int, to_index: int) -> bool:
        if 0 <= from_index < len(self._tracks) and 0 <= to_index < len(self._tracks):
            track = self._tracks.pop(from_index)
            self._tracks.insert(to_index, track)
            return True
        return False
